ion/integration_test.go

```diff
@@ -0,0 +1,807 @@
+	fileBytes := loadFile(t, fp)
+
+	// Make a text writer from the file
+	txtWriter, str := makeTextWriter(t, string(fileBytes))
+
+	// Make a binary writer from the text writer
+	_, buf := makeBinaryWiter(t, []byte(str.String()))
+
+	// Make another text writer using the binary writer
+	txtWriter2, _ := makeTextWriter(t, buf.String())
+
+	//compare the 2 text writers
+	if !reflect.DeepEqual(txtWriter, txtWriter2) {
+		t.Errorf("Round trip test failed on: " + fp)
+	}
+}
+
+// Create a TextWriter from data parameter. Return the writer and string builder containing writer's contents
+func makeTextWriter(t *testing.T, data string) (Writer, strings.Builder) {
+	reader := NewReader(strings.NewReader(data))
+	str := strings.Builder{}
+	txtWriter := NewTextWriter(&str)
+	writeToWriterFromReader(t, reader, txtWriter)
+	txtWriter.Finish()
+	return txtWriter, str
+}
+
+// Create a BinaryWriter from data parameter. Return the writer and buffer containing writer's contents
+func makeBinaryWiter(t *testing.T, data []byte) (Writer, bytes.Buffer) {
+	reader := NewReader(bytes.NewReader(data))
+	buf2 := bytes.Buffer{}
+	binWriter2 := NewBinaryWriter(&buf2)
+	writeToWriterFromReader(t, reader, binWriter2)
+	binWriter2.Finish()
+	return binWriter2, buf2
+}
+
+// Execute loading malformed Ion values into a Reader and validate the Reader.
+func testLoadBad(t *testing.T, fp string) {
+	file, er := os.Open(fp)
+	if er != nil {
+		t.Fatal(er)
+	}
+	defer file.Close()
+
+	r := NewReader(file)
+
+	err := testInvalidReader(t, r)
+
+	if r.Err() == nil && err == nil {
+		t.Fatal("Should have failed loading \"" + fp + "\".")
+	} else {
+		t.Log("expectedly failed loading " + r.Err().Error())
+	}
+}
+
+// Traverse the reader and check if it is an invalid reader, containing malformed Ion values.
+func testInvalidReader(t *testing.T, r Reader) error {
+	for r.Next() {
+		switch r.Type() {
+		case StructType, ListType, SexpType:
+			r.StepIn()
+			testInvalidReader(t, r)
+			r.StepOut()
+		}
+	}
+	if r.Err() != nil {
+		return r.Err()
+	}
+
+	return nil
+}
+
+// Execute equivalency and non-equivalency tests, where true for eq means
+// equivalency and false denotes non-equivalency test.
+func testEquivalency(t *testing.T, fp string, eq bool) {
+	file, er := os.Open(fp)
+	if er != nil {
+		t.Fatal(er)
+	}
+	defer file.Close()
+
+	r := NewReader(file)
+	for r.Next() {
+		embDoc := embeddedDoc(r.Annotations())
+		switch r.Type() {
+		case StructType, ListType, SexpType:
+			var values []item
+			r.StepIn()
+			if embDoc {
+				values = handleEmbeddedDoc(t, r)
+			} else {
+				for r.Next() {
+					values = append(values, readCurrentValue(t, r))
+				}
+			}
+			equivalencyAssertion(t, values, eq)
+			r.StepOut()
+		}
+	}
+	if r.Err() != nil {
+		t.Error()
+	}
+}
+
+// Ion values in top-level sequence annotated with "embedded_documents" parse as Ion string and each
+// of the string values in turn, parse as a document and put in an item struct (defined in this file).
```

Comment:
The `embedded_documents` mechanism is explained pretty succinctly [in the `ion-tests` repo README](https://github.com/amzn/ion-tests). I suggest just copy/pasting it here for clarity:

> If a top-level sequence is annotated with "embedded_documents", it denotes that each of its Ion values is to be parsed as a Ion string value, where its string value is to be parsed as a document. [This function performs that parsing and returns the values in the embedded document as `item`s].

Reply:
I put the explanations in case someone is looking at test and without knowing ```embedded_documents```.
Modified the comment.